interfaces/cython/cantera/composite.py

```diff
@@ -576,6 +595,117 @@ def equilibrate(self, *args, **kwargs):
+        this method allows to restore data exported by `collect_data`.
+
+        :param data: a 2D Numpy array holding data to be restored.
+        :param labels: a list of labels corresponding to SolutionArray entries.
+
+        The receiving SolutionArray either has to be empty or should have
+        matching dimensions. Essential state properties and extra entries
+        are detected automatically whereas stored information of calculated
+        properties is omitted. If the receiving SolutionArray has extra
+        entries already specified, only those will be imported; if *labels* does
+        not contain those entries, an error is raised.
+        """
+
+        # check arguments
+        if not isinstance(data, np.ndarray) or data.ndim != 2:
+            raise TypeError("restore_data only works for 2D ndarrays")
+        elif len(labels) != data.shape[1]:
+            raise ValueError("inconsistent data and label dimensions: "
+                             "({} vs. {})".format(len(labels), data.shape[1]))
+        rows = data.shape[0]
+        if self._shape != (0,) and self._shape != (rows,):
+            raise ValueError('incompatible dimensions.')
+
+        # get full state information (may differ depending on type of ThermoPhase)
+        full_states = self._phase._full_states.values()
+        if isinstance(self._phase, PureFluid):
+            # make sure that potentially non-unique state definitions are checked last
+            last = ['TP', 'TX', 'PX']
+            full_states = [fs for fs in full_states
+                           if fs not in last] + ['TPX'] + last
+
+        # determine whether complete concentration is available (mass or mole)
+        # assumes that `X` or `Y` is always in last place
+        mode = ''
+        has_species = False
+        for prefix in ['X_', 'Y_']:
+            spc = ['{}{}'.format(prefix, s) for s in self.species_names]
+            # solution species names also found in labels
+            valid_species = {s[2:]: labels.index(s) for s in spc
+                             if s in labels}
+            # labels that start with prefix (indicating concentration)
+            all_species = [l for l in labels if l[:2] == prefix]
+            if valid_species:
+                # save species mode and remaining full_state candidates
+                mode = prefix[0]
+                full_states = [v[:-1] for v in full_states if mode in v]
+                break
+        if len(valid_species) != len(all_species):
+            incompatible = list(set(valid_species) ^ set(all_species))
+            raise ValueError('incompatible species information for '
+                             '{}'.format(incompatible))
+        if mode == '':
+            # concentration specifier ('X' or 'Y') is not used
+            full_states = {v[:2] for v in full_states}
+
+        # determine suitable thermo properties for reconstruction
+        basis = {'molar': 'mole', 'mass': 'mass'}[self.basis]
```

Comment:
I'm not sure about this construct, would it be better to say
```
basis = 'mass' if self.basis == 'mass' else 'mole'
```
?

Reply:
:thumbup: ... I routinely use this for longer selections, but your suggestion is clearly better in this instance!